### ISSUE TITLE
Change NotDispatchable keyword into an attribute

### ISSUE THREAD

**Issue by WaynePhillipsEA** [WaynePhillipsEA/twinbasic#806]
In the latest versions of the WinNativeForms package, I'm using the `NotDispatchable` keyword in several places.  This keyword has existed for a very long time in tB, but I don't believe it had been discovered by anyone yet.  It allows for creation of interfaces and classes that don't support IDispatch (i.e. late-binding).  So for example, we can define popular Windows OLE/COM interfaces in tB without the need for external type libraries:

```
[ InterfaceId ("0000000c-0000-0000-C000-000000000046") ]
NotDispatchable Private Interface IStream
...
End Interface
```

Or we can create a class whose implicit default interface doesn't implement IDispatch at all:
```
NotDispatchable Class MyEarlyBindingClass
...
End Class
```

Now that we have an attributes syntax (#25), I want to replace the NotDispatchable keyword with a more appropriate attribute.  In .NET they have the `[ InterfaceTypeAttribute(InterfaceIsIUnknown) ]` attribute, but I'm thinking for us perhaps options of `[ BaseInterface(IDispatch) ]` / `[ BaseInterface(IUnknown) ]` / `[ BaseInterface(Dual) ]`.  With Dual being the default of course.   Any thoughts on the naming/syntax of the attribute?

**Comment by bclothier** [comment]
What if I wanted to do the TLB equivalent of:
```
interface IFoo : IBar
```
? 

Within the context, an interface can only inherit from only one interface, with the `IUnknown` being the ultimate root for any COM interface. In the example, inheriting the `IBar` for the `IFoo` depends on whether `IBar` comes from `IDispatch` or `IUnknown`, so therefore, defining a `BaseInterface` on such interface would not make much sense. 

I believe that currently is not possible in .NET -- they always base it off the `IUnknown`, `IDispatch`, `Dual` or `IInspectable` directly, even if the .NET interface itself inherits from another COM-visible interface. 

To match the full range of possible choices in COM, I'd imagine something like `BaseInterface(Of IFoo)` and `BaseInterface(IInspectable)`*  in addition to the 3 you already proposed.

* `IInspectable` is a WinRT thing. I don't know enough if that is something we want to be able to do in tB.

**Comment by WaynePhillipsEA** [comment]
Interface inheritance would be relatively straight-forward to support, and so yes I agree that it could form part of this proposed change.

So perhaps an attribute is not suitable, and we should instead consider using the Inherits syntax for this.

For interfaces, we shall allow only one inherited interface, and if it is not defined, then `Inherits IDispatchDual` is assumed.  Built-in options of IUnknown, IDispatch and IDispatchDual will be allowed, as well as any other defined interface from the project or any referenced type library.   Specifying `Inherits IDispatch` will create a dispinterface rather than a normal dual interface.

For classes, it's more complicated as we need to think in the future of a time when we will want to support full and proper class inheritance (#73), but initially we shall just restrict it to a single `Inherits` definition, and it must refer to an interface not a class.  This will take effect on the implicit default interface generated by the class.

**Comment by bclothier** [comment]
Ok, let's consider this hypothetical setup:
```
interface IBar : IDispatch {...}
interface IFoo : IBar
interface IBuzz

coclass FooImpl {
  [default] interface IBuzz;
  interface IFoo
}
```

I imagine this would be expressed in tB something like this:
```
Public Interface IBar
...
End Interface

Public Interface IFoo
  Inherits IBar
...
End Interface

Public Interface IBuzz
...
End Interface

[ DefaultInterface(Of IBuzz)]
Public Class FooImpl
  Implements IBuzz
  Inherits IFoo
...
End Class
```

Once we have class inheritance, I guess we either make it a choice between an interface or a class:

```
Public Class FooImpl
  Inherits BarImpl
  Implements IFoo
  Implements IBuzz
End Class
```

or differentiate between the interface inheritance and class inheritance. In C++, we have the flexibility of creating a C++ class that can potentially implement several COM interfaces that are otherwise unrelated and be served by the same class. We should be able to keep the same level of flexibility (complexity? 😓 ) to support everything that is possible in COM / TLB.

**Comment by WaynePhillipsEA** [comment]
```
[ DefaultInterface(Of IBuzz)]
Public Class FooImpl
  Implements IBuzz
  Inherits IFoo
...
End Class
```

Eventually, yes maybe it will look like that.  However, at the moment my main concern in this issue is to remove/replace the NotDispatchable keyword, and I feel this is deviating too much.

```
Public Class FooImpl
  Implements IBuzz
  Inherits IFoo          
...
End Class  
```

So I propose in this example, class FooImpl still generates an implicit _FooImpl interface (as it does now), but _FooImpl inherits IFoo. so:

```
interface _FooImpl : _Foo  { ... }

coclass FooImpl {
  [default] interface _FooImpl;
  interface IBuzz
}
```

This would mean very little change from the current way NotDispatchable is implemented, when `Inherits IUnknown` is specified.

**Comment by WaynePhillipsEA** [comment]
Perhaps it would be more clear if the Inherits syntax was part of the Interface / Class syntax.  eg.

```
Public Class FooImpl Inherits IFoo
  Implements IBuzz
...
End Class  
```

That way, we can later differentiate between the inheritance of the implicit default interface, and proper class inheritance.  Or perhaps even `Inherits Interface IFoo` to be absolutely clear.

**Comment by WaynePhillipsEA** [comment]
After some further thought, I'm inclined to only implement the `Inherits` syntax for interfaces for now.  This will at least allow me to remove all public uses of the NotDispatchable keyword in the built in packages.  

We can then later discuss how best to control the class implicit interface inheritance, perhaps after full inheritance support is included in tB.

**Comment by wqweto** [comment]
IMO it’s completely unnecessary to support disponly interfaces i.e. in 2022 only two options make sense - vtbl or dual. Dispinterfaces can be 100% replaced by dual ones.

Also inheriting from IBar is completely different ball game as the compiler cannot auto-implement its methods as it does for base IDispatch/IUnknown so no need to mix topics IMO.

**Comment by WaynePhillipsEA** [comment]
> IMO it’s completely unnecessary to support disponly interfaces i.e. in 2022 only two options make sense - vtbl or dual. Dispinterfaces can be 100% replaced by dual ones.

Not if we're defining a dispinterface that is outside of our control (e.g. defining an existing dispinterface inside a tB twin file instead of importing it from a type library).

> Also inheriting from IBar is completely different ball game as the compiler cannot auto-implement its methods as it does for base IDispatch/IUnknown so no need to mix topics IMO.

Again this would be for recreating interfaces within tB without an external type library.  So for example, we'd be able to define IStream properly so that it inherits from ISequentialStream that inherits from IUnknown.

Not only do I want tB to be the best COM consumer, I also want it to be the best COM producer as well, so supporting as much of the COM standards as possible is one of the more general aims of this project.

**Comment by wqweto** [comment]
I see your point but I’ve not seen anyone declaring and/or producing disponly interfaces (outside of VBx events support which is baked into TB already) as dual interfaces supersede these in every regard. If a language/IDE supports dispinterfaces only it still can be fed dual interfaces instead. Anyway, I see your point supporting these for the sake of having complete support of all base kinds.

**Comment by bclothier** [comment]
> However, at the moment my main concern in this issue is to remove/replace the NotDispatchable keyword, and I feel this is deviating too much from my objective.

Yes, I'm mixing the issues, mainly because I wanted to be sure we don't end up painting ourselves into a corner when the time comes to implement the true class inheritance.

After sleeping / thinking about it some more, I would propose that anything that pertains to the type library generation should be consistently managed by attributes and true class inheritance be a language feature. As was [commented in #73](https://github.com/WaynePhillipsEA/twinbasic/issues/73#issuecomment-822123488), what inheritance we expose via the type library is separate from what inheritance we choose to implement within the code itself. 

By using attributes we can maintain the separation between the shape of the type library and the shape of the classes. I think that will give us more flexibility and more importantly, the freedom to change the internal implementation without changing the type library, which will be very important in being able to version cleanly without breaking the published interfaces. 

For that reason, I think treating the type library generation as a compiler service rather than a language feature will be cleaner.

**Comment by WaynePhillipsEA** [comment]
@Greedquest I like that!

**Comment by Kr00l** [comment]
In UnrealScript it was used to be `expands` keyword. Nowadays it allows either `extends` or `expands`.

`class classname extends parentclass`

Just wanted to throw that out. It was my first experience with coding in the map editor since then.. ;D

**Comment by bclothier** [comment]
I wanted to ask one more question regarding the interfaces. In .NET world, it is possible to do:
```
interface IFizzBuzz : IFizz, IBuzz
```

which is convenient for composition and helps a lot with authoring an implementation that needs to satisfy multiple requirements without creating a giant one-size-fits-all interface. Compositions are typically done using interfaces, rather than classes.

However, that won't fly for the type library's purpose --- all interfaces defined in the type library can only have one parent, and `IUnknown` better be the greatest granddaddy of them all (or else!). 

So the question is --- how do we promote composition in tB without being tied to the type library's restrictions? IMPOV, those 2 issues are separate and I don't want to overlap them but the proposed `Inherits` would prevent creating a composed interface? 

One way to avoid this is to have tB implement mixins/traits as an alternative to composition, which was discussed in #73 but I don't know enough if that is a necessarily a good idea or not.

**Comment by WaynePhillipsEA** [comment]
In BETA 32, the `Extends` syntax is now available on interfaces:

```
[ InterfaceId ("0c733a30-2a1c-11ce-ade5-00aa0044773d") ]
Private Interface ISequentialStream Extends stdole.IUnknown
    Sub Read(ByVal pv As LongPtr, ByVal cb As Long, ByRef pcbRead As Long)
    Sub Write(ByVal pv As LongPtr, ByVal cb As Long, ByRef pcbWritten As Long)
End Interface

[ InterfaceId ("0000000c-0000-0000-C000-000000000046") ]
Private Interface IStream Extends ISequentialStream
   ...
End Interface
```

If not specified, interfaces extend from IDispatch.  There are a couple of known issues that will be addressed over the next few days:
- intellisense doesn't show the base interface members, despite them being available and working
- the combobox dropdowns above the code editor are also not showing the base interface members

**Comment by bclothier** [comment]
Some possibly cosmetic notes:
![image](https://user-images.githubusercontent.com/2367644/167698534-a330538c-51b6-4320-bcb0-563367883bcf.png)

1) The interface is called "class"?
2) The interface has interfaces, too? Or is it a matter of a spurious adding of underscore letter?
3) Not sure if `IUnknown` on the last line is redundant since we already have the map showing it's extending from the `IUnknown`. 

To confirm, if the class implements a derived interface, does it need to implement the parent interface explicitly to be castable to that parent interface or is that implied? 

Also, any thoughts on the interface composition question?

**Comment by WaynePhillipsEA** [comment]
Yeah there's a bit of tech debt there.  You may recall that when I first added `Interface` support, I _temporarily_ bodged it together by treating them as "special" classes internally... Needless to say that they are still "special" classes internally!  That is on my long list to sort out one day ;)

The final IUnknown is the "identity" IUnknown that we add to classes.  VBx does similar.

So this is mostly cosmetic as you just shouldn't be seeing the psuedo class at all.  I'll get that fixed.

**Comment by WaynePhillipsEA** [comment]
> To confirm, if the class implements a derived interface, does it need to implement the parent interface explicitly to be castable to that parent interface or is that implied?

You'd need to implement the outermost interface, and the rest would be implied.

> Also, any thoughts on the interface composition question?

That one is quite tricky.  It's probably something to look at for v2.

**Comment by WaynePhillipsEA** [comment]
BTW the reason I implemented this feature now was because the WebView2 interfaces rely heavily on interface inheritance.  Check out the new WebView2Package....    see ICoreWebView2, ICoreWebView2_2, ICoreWebView2_3, ICoreWebView2_4, ICoreWebView2_5, ICoreWebView2_6, ICoreWebView2_7, ICoreWebView2_8, ICoreWebView2_9...

**Comment by bclothier** [comment]
Need more interfaces.

**Comment by jpbro** [comment]
Must have interfaces 🧟

Just a side note - it might just be me, but I find all the [ ] stuff in code very distracting. This is probably better as a new discussion topic, but I wonder if it would be possible to either:

A) Have a switch for toggling the display of [ ] block stuff (Attribute, InterfaceID, etc...), or 

B) Have an option to show all [ ] stuff off to the side in a pane in the IDE as opposed to in the code window? E.g. when you move the caret into a new method, the pane would show the attributes, interfaceIDs, etc.. that are connected to that method.

**Comment by WaynePhillipsEA** [comment]
@jpbro we could possibly use folding ranges in the Monaco editor to hide them, and then they can be expanded when necessary.   I'd have to experiment with that.   

That said, I don't personally find them distracting, in the dark theme anyway.  But I take your point.

**Comment by jpbro** [comment]
@WaynePhillipsEA It could just be me, and it's certainly nothing critical for me either. I'm sure I'll get used to ignoring those elements after a while.

That said, a visibility toggle might be nice anyway if it were possible, and while I'm thinking of it - a visibility toggle for comments would be nice too. Sometimes you just want to stare at code and everything else kind of gets in the way (especially things like long comment runs in between a few lines of code)

**Comment by WaynePhillipsEA** [comment]
@jpbro no problem!  I doubt you're the only one, so it's definitely worth consideration.  Thanks.

**Comment by WaynePhillipsEA** [comment]
> Some possibly cosmetic notes:

@bclothier that should now be fixed in BETA 33.  Thanks.

**Comment by WaynePhillipsEA** [comment]
The two issues with the Extends functionality mentioned earlier in this thread, have now been resolved in BETA 45:
-  intellisense doesn't show the base interface members, despite them being available and working
-  the combobox dropdowns above the code editor are also not showing the base interface members
